### ISSUE TITLE
Session Replay support for Flutter on Mobile

### ISSUE THREAD

**Issue by bruno-garcia** [getsentry/sentry-dart#1193]
Add support for Sentry's Session Replay: https://sentry.io/for/session-replay/

**Comment by Rafik-Belkadi-Reccap** [comment]
from how many thumbs up will you start to consider its implementation?

**Comment by marandaneto** [comment]
> from how many thumbs up will you start to consider its implementation?

This is blocked by https://github.com/flutter/flutter/issues/117382

**Comment by YasserDRIF** [comment]
We are using Smartlook currently, Adding this to Sentry would be amazing

**Comment by bruno-garcia** [comment]
We're working on it! Wanna join the early adopter release? Join the waitlist and discussion about the feature: 
* https://github.com/getsentry/sentry/discussions/63138

**Comment by vaind** [comment]
Replay alpha version now available for Android in 8.6.0-alpha.2 - please share any and all feedback.


  To try out replay, you can set following options:

  ```dart
  await SentryFlutter.init(
    (options) {
      ...
      options.experimental.replay.sessionSampleRate = 1.0;
      options.experimental.replay.errorSampleRate = 1.0;
    },
    appRunner: () => runApp(MyApp()),
  );
  ```

  Access is limited to early access orgs on Sentry. If you're interested, [sign up for the waitlist](https://sentry.io/lp/mobile-replay-beta/)

**Comment by mcosti** [comment]
Where should we post the issues we find? In this topic or each in a separate issue?

Here's what I found so far:

```
[sentry] [debug] WidgetFilter obscuring: Text("Editor")
[sentry] [debug] WidgetFilter obscuring: Text("My Festivals")
[sentry] [debug] WidgetFilter obscuring: Text("Messages")
[sentry] [debug] WidgetFilter obscuring: Text("Profile")
[sentry] [error] Replay: failed to capture screenshot.
         'dart:ui/painting.dart': Failed assertion: line 26 pos 10: '<optimized out>': Rect argument contained a NaN value.
         #0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:51:61)
         #1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:40:5)
         #2      _rectIsValid (dart:ui/painting.dart:26:10)
         #3      _NativeCanvas.drawRect (dart:ui/painting.dart:5936:12)
         #4      ScreenshotRecorder._obscureWidgets (package:sentry_flutter/src/replay/recorder.dart:137:14)
         #5      ScreenshotRecorder._capture (package:sentry_flutter/src/replay/recorder.dart:103:9)
         <asynchronous suspension>
         #6      Scheduler._run.<anonymous closure> (package:sentry_flutter/src/replay/scheduler.dart:53:41)
         <asynchronous suspension>
```


Also, quite low performance, especially on Android. I am developing a chat app and I can see a ton of logs for obscuring data, including the bottom navigation which gets obscured on every scroll. (Although not technically re-rendered).


Edit: I do not see any recordings showing up for iOS and for Android, the whole 9 minute recording is stuck on the initial first tab (not the loading screen, but the first screen from the navigation bar). I can see the network logs as I'm scrolling through my app, but the recording itself is stuck. 


[✓] Flutter (Channel stable, 3.22.3, on macOS 14.1 23B74 darwin-arm64, locale en-NL)

**Comment by vaind** [comment]
> I do not see any recordings showing up

I suspect that's because of the error - it fails during frame creation so the previous frame gets repeated. I'll have a look at the failing assertion.

If the app you're working on is available publicly (and easy to set up), I can have a look which components are causing the issue.

**Comment by mcosti** [comment]
My app is not available publicly, but I am more than willing to assist with anything I can help with.

I've just had a look and the error gets triggered on every page I have, both including and excluding pages that have the bottom navigation bar (that was a hunch I had, but it's not that)

**Comment by mcosti** [comment]
Replacing the code with this:

```dart
  void _obscureWidgets(Canvas canvas, List<WidgetFilterItem> items) {
    final paint = Paint()..style = PaintingStyle.fill;
    for (var item in items) {
      paint.color = item.color;
      if (item.bounds.hasNaN) {
        _logger(SentryLevel.debug,
            "Replay: skipping widget with NaN bounds: $item. ${item.bounds}");
        continue;
      }
      canvas.drawRect(item.bounds, paint);
    }
  }
```

Works, of course.


Adding this inside of _obscureIfNeeded:

```dart
    if (rect.hasNaN) {
      logger(SentryLevel.debug, "Widget $widget has NaN Bounds. $offset $size");
    }
```

Results in these logs:

```
[sentry] [debug] Widget Text("Contact", debugLabel: ((englishLike bodyLarge 2021).merge((((whiteMountainView bodyLarge).apply).apply).merge((((whiteMountainView bodyLarge).apply).apply).merge(unknown)))).apply, inherit: false, color: Color(0xffffffff), family: Poppins_regular, familyFallback: [Poppins], size: 14.0, weight: 400, letterSpacing: 0.5, baseline: alphabetic, height: 1.5x, leadingDistribution: even, decoration: Color(0xffececec) TextDecoration.none) has NaN Bounds. Offset(NaN, NaN) Size(229.4, 21.0)
```



I should mention that the Contact label is not visible at all in the moment of running this code.


To give some more hints about how my app works,

```dart
final PageController navigationPageController = PageController();
...



  Widget build(BuildContext context) {
    WidgetsBinding.instance.addPostFrameCallback((_) => checkIfOnboardingShouldBeShown(context, mounted));
    return Scaffold(
        bottomNavigationBar: ValueListenableBuilder(
            valueListenable: currentNavigationIndex,
            builder: (context, currentIndexValue, child) {
              return Theme(
                data: ThemeData(
                  splashColor: Colors.transparent,
                  highlightColor: Colors.transparent,
                  brightness: Theme.of(context).brightness,
                ),
                child: BottomNavigationBar(
                    onTap: (index) => navigationPageController.jumpToPage(index),
                    currentIndex: currentIndexValue,
                    type: BottomNavigationBarType.fixed,
                    items: [
                      BottomNavigationBarItem(icon: const Icon(Icons.cut), label: tr(LocaleKeys.app_editor)),
                      BottomNavigationBarItem(
                          icon: const Icon(FontAwesomeIcons.calendar), label: tr(LocaleKeys.app_myFestivals)),
                      const BottomNavigationBarItem(icon: Icon(Icons.chat_sharp), label: 'Messages'),
                      BottomNavigationBarItem(icon: const Icon(Icons.manage_accounts), label: tr(LocaleKeys.app_me))
                    ]),
              );
            }),
        body: UpgradeAlert(
          upgrader: upgrader,
          child: PageView(
            controller: navigationPageController,
            onPageChanged: (index) => currentNavigationIndex.value = index,
            children: [VideoEditorHomePage(), MyFestivalsScreen(), ChatScreen(), ProfileScreen()],
          ),
        ));
```


I am using a PageView

**Comment by vaind** [comment]
> I am developing a chat app and I can see a ton of logs for obscuring data, including the bottom navigation which gets obscured on every scroll. (Although not technically re-rendered).

and 

> I should mention that the Contact label is not visible at all in the moment of running this code.

Understanding which widgets are visible and which are obscured/overlayed by something else can be tricky. We have some logic that checks out-of-screen widgets as well as some common ways to hide them (visibility/opacity). I'm open to suggestions how this could be done, at least in your scenario - we can see if it can be generalized or there are counter-examples. 

That the widget filter evaluates the `Contact` label means it is part of the widget tree and it hasn't been trimmed with the visibility heuristics. Any chance you can show the widget tree screenshot from developer tools so we can see if there's anything we can rely on (generically) to trim out hidden pages?

**Comment by mcosti** [comment]
I should mention that I am a python developer, so this kind of thing is completely out of my realm, but I'll do my best.

This is the widget tree:



![Image](https://github.com/user-attachments/assets/798e92b4-159d-4564-8a48-1a281ef8748d)


As you can see, all the pages are there.



When the contact button is not in view
![Image](https://github.com/user-attachments/assets/29619ba8-a864-44c9-8cc2-55d837fe37ac)

When the contact button is in view


![Image](https://github.com/user-attachments/assets/06166daf-7389-4410-b86c-13deb0f51710)


The only difference I could see is "Needs repaint"

**Comment by vaind** [comment]
> The only difference I could see is "Needs repaint"

Yes, I can't see anything else either. I'm looking for some component that decides whether it's visible. Maybe the itself has visibility or sth 🤔

**Comment by vaind** [comment]
@mcosti I've tried to reproduce what you're seeing with `NaN` with `PageView` but couldn't. Any chance you can come up with a repro that you're able to share?

**Comment by vaind** [comment]
TODOs

* [x] (blocker) Provide a way to redact specific elements (hook into redaction logic) https://github.com/getsentry/sentry-dart/pull/2324
* [ ] (blocker) Mask custom graphics https://github.com/getsentry/sentry-dart/pull/2346
* [ ] (blocker) Mask webviews - https://github.com/getsentry/sentry-dart/pull/2355
* [ ] (blocker) Mask video - https://github.com/getsentry/sentry-dart/pull/2355
* [x] (blocker) Measure overhead
* [x] (blocker) https://github.com/getsentry/sentry-dart/pull/2349
* [x] (blocker) https://github.com/getsentry/sentry-docs/issues/11555
* [ ] Pause/resume API 
* [x] support orientation change - seems to work out of the box via native integration
* [ ] Clock drift fix: use UTC, or single time clock read, or stopwatch
* [x] Add quality setting low/medium/high 
* [x] report SDK sample rates

**Comment by rhinck** [comment]
I apologize for my ignorance, but does this issue being closed indicate that there is now Flutter support for session replay? If so, how would I go about trying it out in my project? (I couldn't seem to find any docs offering additional information)

**Comment by vaind** [comment]
Session Replay beta for iOS and Android should come out in the next release

**Comment by buenaflor** [comment]
It's available starting from Sentry Flutter `8.9.0`

**Comment by bruno-garcia** [comment]
@vaind could you help get it on our docs? https://docs.sentry.io/platforms/flutter/

If we add a note that it's experimental, like we did for the other SDKs, it should be OK. Now is the time since we expect to GA this in a a month or two.

**Comment by darkfrog26** [comment]
> Session Replay beta for iOS and Android should come out in the next release

I presume this means that it's not supported for Flutter Web?

**Comment by buenaflor** [comment]
> I presume this means that it's not supported for Flutter Web?

~It's not available on Flutter Web, at least not yet. I have a working experimental/draft [implementation](https://github.com/getsentry/sentry-dart/pull/2238) but I can't give a sure timeline when this will be available~

The draft implementation is not sufficient due to PII

**Comment by darkfrog26** [comment]
@buenaflor thanks for the info, I appreciate it.

**Comment by ant1m4tt3r** [comment]
if we use this experimental feature would we be able to see the replays in my sentry dashboard? looks like it still isn't available i guess. am i missing something?

![Image](https://github.com/user-attachments/assets/02e06f7a-006e-43bd-8783-cf06012d53b7)

**Comment by bruno-garcia** [comment]
> if we use this experimental feature would we be able to see the replays in my sentry dashboard? looks like it still isn't available i guess. am i missing something?
> 
> ![Image](https://github.com/user-attachments/assets/02e06f7a-006e-43bd-8783-cf06012d53b7)

Yeah we just haven't updated Sentry to consider Flutter as supported. But if you send the data, it'll show up.

**Comment by ant1m4tt3r** [comment]
thanks @bruno-garcia. for some reason i can't see the data yet, even though the app is running with the replay feature on. i'm looking into it, it may be some misconfiguration or something that i am missing.

should the replay work in a debug build or only in release mode?

**Comment by vaind** [comment]
> should the replay work in a debug build or only in release mode?

Works in debug build too. You can try turning on debug logging in options to get more info on what's going on.

**Comment by pixnbit** [comment]
Is there a waitlist to sign up for? I can't find where to sign up, I would love try out this feature. I think it's a game changer

**Comment by bruno-garcia** [comment]
> thanks [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/1633368?s=64&amp;u=856a9e52d4cdc92eac2ff9e1f97900df39ae5e4d&amp;v=4">@bruno-garcia](https://github.com/bruno-garcia). for some reason i can't see the data yet, even though the app is running with the replay feature on. i'm looking into it, it may be some misconfiguration or something that i am missing.
> 
> should the replay work in a debug build or only in release mode?

Are you a member of an enterprise org on Sentry SaaS? That's the only thing I can think of. Our open beta is only available to self-serve orgs right now.

If that's the case, can you create a free developer account, and try there, does it work?

**Comment by coffeeispower** [comment]
> > thanks [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/1633368?s=64&amp;u=856a9e52d4cdc92eac2ff9e1f97900df39ae5e4d&amp;v=4">@bruno-garcia](https://github.com/bruno-garcia). for some reason i can't see the data yet, even though the app is running with the replay feature on. i'm looking into it, it may be some misconfiguration or something that i am missing.
> > should the replay work in a debug build or only in release mode?
> 
> Are you a member of an enterprise org on Sentry SaaS? That's the only thing I can think of. Our open beta is only available to self-serve orgs right now.
> 
> If that's the case, can you create a free developer account, and try there, does it work?

I'm also trying to use this feature but it just refuses to work, it doesn't send the replay or at least sentry is not accepting it. I just created the sentry account for the project, it's a developer plan org.

**Comment by nrajput05** [comment]
> > I presume this means that it's not supported for Flutter Web?
> 
> It's not available on Flutter Web, at least not yet. I have a working experimental/draft [implementation](https://github.com/getsentry/sentry-dart/pull/2238) but I can't give a sure timeline when this will be available

Is this still not available for flutter web ?

**Comment by buenaflor** [comment]
> Is this still not available for flutter web ?

no, as I said above I can't give a sure timeline when this will be available, it could still take a long time until we properly tackle this. I will write it here if I have any further information/updates

**Comment by skiunke** [comment]
> > Is this still not available for flutter web ?
> 
> no, as I said above I can't give a sure timeline when this will be available. I will write it here if I have any further information/updates

would a potential implementation cover WASM support?

**Comment by buenaflor** [comment]
@skiunke I have to check but my hunch is yes

**Comment by bruno-garcia** [comment]
> > > Is this still not available for flutter web ?
> > 
> > 
> > no, as I said above I can't give a sure timeline when this will be available. I will write it here if I have any further information/updates
> 
> would a potential implementation cover WASM support?

I updated the title of the issue to be clear but this support is all about Android and iOS at this time. While our JS SDK could record a replay from Flutter Web, it would be based off of our web canvas support, have no masking at all and quite high overhead. So it would require a new approach for Flutter Web which we didn't look into yet.
